enhancements/machine-api/control-plane.md

```diff
@@ -0,0 +1,281 @@
+6 - Pick oldest machine in more populated domain failure.
+7 - Trigger scale out workflow (starting in 4). Set "replaced" annotation. Requeue.
+This is effectively a rolling upgrade with maxUnavailable 0 and maxSurge 1.
+
+#### Self healing (MHC + reconciling)
+1 - The Controller should always reconcile by removing Etcd members for voluntary Machine disruptions, i.e machine deletion.
+2 - At creation time, unless indicated otherwise `EnableAutorepair=true` will be default. The controller will create a Machine Health Checking resource with an ownerRef which will monitor the Control Plane Machines. This will request unhealthy machines to be deleted. Related https://github.com/openshift/machine-api-operator/pull/543.
+3 - The controller will ensure the `maxUnhealthy` value in the MHC resource is set a to known integer to prevent farther remediation from happening during scenarios where quorum could be violated. E.g 1 for cluster with 3 members or 2 for a cluster with 5 members.
+
+#### Bootstrapping (scale out)
+Currently during a regular IPI bootstrapping process the installer uses Terraform to create a bootstrapping instance and 3 master instances. Then it creates Machine resources to "adopt" the existing master instances. In the past Etcd quorum needed to be reached between the three of them before having storage available for the control plane to run self hosted and so for the CVO to run its payload.
+
+The Cluster Etcd Operator introduced support for a single member Etcd cluster to be available quickly on the bootstrapping machine. This lets the CVO to be deployed much faster while new Etcd members are added organically as masters come up.
+
+This proposes dropping terraform for creating masters instances in favour of letting the installer to define a Control Plane resource that scales from zero to 3 replicas as soon as the CVO runs the Machine API Operator. Alternatively if this happen to not be doable because of chicken-egg issues, we could keep the current workflow and include an additional step to create the Control Plane resource which would just adopt existing Master machines.
+
+#### API
+```
+type ControlPlane struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   ControlPlaneSpec   `json:"spec,omitempty"`
+	Status ControlPlaneStatus `json:"status,omitempty"`
+}
+
+type ControlPlaneSpec struct {
+	// Number of desired machines. Defaults to 1.
+	// Only odd numbers are permitted.
+	Replicas *int32 `json:"replicas,omitempty"`
+
+	// EnableAutorepair defines if a Machine Health Checking resource
+	// should be created.
+	// This will autorepair faulty machine/nodes in scenarios where quorum would not be violated.
+	// e.g 1 unhealthy out of 3.
+	// Defaults to true.
+	EnableAutorepair bool `json:"enableautorepair,omitempty"`
+	
+	// ProviderSpec details Provider-specific configuration to use during node creation.
+	ProviderSpec ProviderSpec `json:"providerSpec"`
```

Comment:
+1.

What is providerSpec in this context?

My thoughts were rather than try to dynamically build machine's from templates and several data sources that are different per-provider, the installer can populate a list of templates for each supported AZ at creation time.  Basically a machineset, but the controller chooses 1 of N templates rather than just the one.